eclipse-theia/theia
# Support vscode TypeHierarchy API (references-view)

#### What it does


Fixes: #11516 

This change introduces support for vscode's `Typehierarchy` API and integrates the solution with the facilities provided by the `references-view` extension.

E.g. via the implementation of the commands 
- `vscode.prepareTypeHierarchy`,
- `vscode.provideSupertypes`,
- `vscode.provideSupertypes`

The `Show Type Hierarchy` context menu shall be enabled from an editor as long as a `TypeHIerarchyProvider` is registered for the specific language.

Similarly the following commands shall now be functional from the command palette.
- `Types: Show Type Hierarchy`
- `Types: Show Subtypes`
- `Types: Show Supertypes`



#### How to test

1. Download the following test extension:
https://github.com/microsoft/vscode/files/7033356/typeh-0.0.1.vsix.zip
2. Correcting a bug in the extension (I am referencing the original work, rather than copying it)
Open the zip file and change LINE: 30 to: 
>  if (offset <= position.character && position.character <= offset + word.length) {
i.e. from `<` to `<=`
3. Modify the file name to have the extension `.vsix`
4. Start `theia` and install the test extension 
5. The test extension registers type providers for plain text files i.e. `.text`, and provides `subtypes` corresponding to each word of the next line, it provides `supertypes` corresponding to each word of the previous line.
In a `workspace`, create a `text` file with multiple lines of text on it.
6. From the editor, select a word in your document, open the context menu and select `Show Type Hierarchy`
7. Verify that the references view gets populated with words from the line below as `subtypes` 
8. Click the button on the references view to toggle and see the `supertypes` of the initial selection.
9. From the command palette verify that the following commands work similarly to the above.
- `Types: Show Type Hierarchy`
- `Types: Show Subtypes`
- `Types: Show Supertypes`
10. Make sure `Call hierarchy` works as before

https://user-images.githubusercontent.com/76971376/191809730-d7af7682-032a-455f-903f-42215129db43.mp4

NOTE: The functionality can be tested using `clangd` although the feature is recently introduced and would need the following:
- Make sure you have the native `clangd` at version 15.0.0 or higher
- Stub the `vscode` API for `Inlay hints` in `theia`
- Move `theia`'s  DEFAULT_SUPPORTED_API_VERSION to `1.67.0'`
- Install the `Clangd` extension (id: `llvm-vs-code-extensions.vscode-clangd`)
- Make sure the following preference points to the location of your native `clangd` executable: `Clangd: Path`

Given the above, the test extension is a more practical alternative :-)

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
